libs/wingc/src/jsify.rs

```diff
@@ -1148,3 +1167,175 @@ fn is_mutable_collection(expression: &Expr) -> bool {
+			}
+		};
+
+		if s.name != "this" {
+			return false
+		}
+
+		// resolve the type of the resource by looking up "this" in our environment
+		let resource_type = self.env
+			.lookup(s, None).expect("'this' to be found")
+			.as_variable().expect("'this' to be a variable")
+			.type_;
+
+		let r = match &*resource_type {
+			Type::Resource(r) => r,
+			_ => panic!("'this' is not a resource"),
+		};
+
+		// lookup our field in the class environment
+		let field_kind = r.env
+			.lookup(property, None).expect("unable to find field in resource env")
+			.as_variable().expect("field is not a variable");
+
+		// we only care about preflight fields (inflight fields are normal references)
+		if field_kind.flight == Phase::Preflight {
+
+			// don't allow capturing reassignable (`var`) fields.
+			if field_kind.reassignable {
+				self.diagnostics.push(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!(
+						"Unable to reference \"this.{}\" from inflight method \"{}\" because it is reassignable (\"var\")",
+						property.name, 
+						self.method_name.name,
+					),
+					span: Some(property.span.clone()),
+				});
+
+				return true
+			}
+
+			// check if the type is capturable (resource, primitive or immutable collection)
+			if ! field_kind.type_.is_capturable() {
+				self.diagnostics.push(Diagnostic {
+					level: DiagnosticLevel::Error,
+					message: format!(
+						"Unable to reference \"this.{}\" from inflight method \"{}\" because type {} is not capturable",
+						property.name, 
+						self.method_name.name,
+						field_kind.type_,
+					),
+					span: Some(property.span.clone()),
+				});
+
+				return true
+			}
+
+			debug!("field \"this.{}\" is referenced from inflight method {}", property.name, self.method_name.name);
+
+			// get/create an "ops" set for this field
+			let ops = self.references.entry(property.name.clone()).or_default();
+
+			if ! self.path.is_empty() {
+				let op = self.path[0].clone();
+
+				// if we are referencing a resource, verify that the operation is an inflight method
+				if let Some(r) = field_kind.type_.as_resource() {
+					if !has_inflight_method(r, &op) {
+						self.diagnostics.push(Diagnostic {
+							level: DiagnosticLevel::Error,
+							message: format!(
+								"Unable to reference \"{}\" from inflight method \"{}\" because it is not an inflight method",
+								self.path.join("."), 
+								self.method_name.name,
+							),
+							span: Some(property.span.clone()),
+						});
+
+						return true
+					}
+				}
+
+				ops.insert(op);
+			}
+
+			self.path.clear();
+
+			// no need to visit recursively
+			return true
+		}
+
+		return false
+	}
+}
+
+/// Checks if a resource class has an inflight method with the given name
+fn has_inflight_method(resource_class: &Class, method_name: &String) -> bool {
+	if let Some(method_var) = resource_class.get_method(method_name) {
+		method_var.flight == Phase::Inflight && !method_var.is_static
+	} else {
+		false
+	}
+}
```

Comment:
nit: I don't think this needs to be a free floating function. Also it's more idiomatic to use `&str` than `&String`, since the latter requires the string to be heap-allocated:
```suggestion
impl Class {
	/// Checks if a resource class has an inflight method with the given name
	fn has_inflight_method(method_name: &str) -> bool {
		// ...
	}
}
```

Reply:
I feel adding `has_inflight_method` to `Class` is too specific. I'll refactor and push a cleanup.

Reply:
I inlined this condition